util/timer/src/timer.rs

```diff
@@ -0,0 +1,254 @@
+use parking_lot::{Condvar, Mutex, RwLock};
+use std::cmp::Reverse;
+use std::collections::binary_heap::BinaryHeap;
+use std::collections::hash_map::{Entry, HashMap};
+use std::collections::VecDeque;
+use std::string::ToString;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{channel, Receiver, RecvTimeoutError, Sender};
+use std::sync::Arc;
+use std::thread::{self, JoinHandle};
+use std::time::{Duration, Instant};
+
+pub type TimerId = &'static str;
+pub type TimerToken = usize;
+
+pub trait TimeoutHandler {
+    fn on_timeout(&self, timer_token: TimerToken);
+}
+
+type TimeoutHandlerMap = HashMap<TimerId, Arc<TimeoutHandler + Send + Sync>>;
+
+pub struct TimerLoop {
+    timers: Arc<RwLock<TimeoutHandlerMap>>,
+    tx: Sender<TimerMessage>,
+    waiter_loop: JoinHandle<()>,
+    workers: Workers,
+}
+
+impl TimerLoop {
+    pub fn new(worker_size: usize) -> TimerLoop {
+        let (tx, rx) = channel();
+        let timers = Arc::new(RwLock::new(HashMap::new()));
+
+        let workers = Workers::new(worker_size, Arc::clone(&timers));
+        let worker_queue = workers.queue.clone();
+
+        let waiter_loop = {
+            thread::Builder::new()
+                .name("Timer waiter".to_string())
+                .spawn(move || waiter_loop(rx, worker_queue))
+                .unwrap()
+        };
+
+        TimerLoop {
+            tx,
+            waiter_loop,
+            timers,
+            workers,
+        }
+    }
+
+    pub fn stop(self) {
+        self.workers.stop();
+        self.tx.send(TimerMessage::Stop).unwrap();
+        self.waiter_loop.join().unwrap();
+    }
+
+    pub fn new_timer<T>(&self, id: TimerId, handler: Arc<T>) -> TimerApi
+    where
+        T: 'static + TimeoutHandler + Send + Sync, {
+        let mut timers = self.timers.write();
+        match timers.entry(id) {
+            Entry::Occupied(_) => panic!("Timer id was already occupied: {}", id),
+            Entry::Vacant(entry) => {
+                entry.insert(handler);
+                return TimerApi {
+                    timer_id: id,
+                    tx: self.tx.clone(),
+                }
+            }
+        };
+    }
+}
+
+pub struct TimerApi {
+    timer_id: TimerId,
+    tx: Sender<TimerMessage>,
+}
+
+impl TimerApi {
+    pub fn schedule(&self, after: Duration, timer_token: TimerToken) {
+        self.tx
+            .send(TimerMessage::Schedule(TimerSchedule {
+                at: Instant::now() + after,
+                timer_id: self.timer_id,
+                timer_token,
+            }))
+            .unwrap();
+    }
+}
+
+struct TimerSchedule {
+    at: Instant,
+    timer_id: TimerId,
+    timer_token: TimerToken,
+}
+
+impl std::cmp::PartialEq for TimerSchedule {
+    fn eq(&self, other: &TimerSchedule) -> bool {
+        self.at.eq(&other.at)
```

Comment:
Is it intentional that two TimerSchedules are treated identical even the timer_id and timer_token is different?

Reply:
It was intenttional. If you want it more explicit, how about introducing a wrapper struct like ```std::cmp::Reverse```?
```rust
struct TimeOrdered<T>(T); 
```
And what about introducing FractionalOrd Trait like this?
```rust
trait FractionalOrd {
    type OrderingFraction: Ord;
    fn ordering_fraction(&self) -> A;
}
```
and blanket impl for ```Ord```, ```PartialOrd```, ```PartialEq``` ?
```rust
impl<T> std::cmp::PartialEq for T
    where T: FractionalOrd
{
    fn eq(&self, other: &T) -> bool {
        let a = self.ordering_fraction();
        let b = other.ordering_fraction();
        a.eq(b)
    }
}
```
Or I could simply make it plain.
```rust
struct TimeOrdered(TimerSchedule); 
impl std::cmp::PartialEq for TimeOrdered { ... }
```

Reply:
Suggested blanket impl conflicts with existing blanket impl 
```
impl<'a, 'b, A, B> PartialOrd<&'b B> for &'a A 
```